### PR TITLE
[Feature](Nereids)support insert into select.

### PR DESCRIPTION
**Pull request by sohardforaname**
# Proposed changes

Issue Number: close #xxx

support insert the ret-value of a query into a table with `partition`, `with label`, `cols` tags:
Like：
```sql
insert into t partition (p1, p2) with label label_1 (c1, c2, c3) [hint1, hint2] with cte as (select * from src) select k1, k2, k3 from cte
```

we create new class: InsertIntoTableCommand, Unbound/Logical/PhysicalOlapTableSink to describe the command of insert and the olapTableSink for Nereids. We make UnboundOlapTableSink in parsing phase and bind it, then implement and translate the node to OlapTableSink. Then we run the command with a transaction.

Describe your changes.

## Checklist(Required)

* [ ] Does it affect the original behavior
* [x] Has unit tests been added
* [ ] Has document been added or modified
* [ ] Does it need to update dependencies
* [x] Is this PR support rollback (If NO, please explain WHY)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by 924060929** [comment]
A known bug: insert into unique table without group by unique key will cause the replication different.

3 replications maybe receive the data with different when exchange.

e.g.
```
+-----------+                        +--------------------------+ 
| machine 1 |        exchange        |   machine 4 (replica 1)  |
+-----------+                        +--------------------------+
|   'a'     |                        |       'a', 'b', 'c'      |
+-----------+                        +--------------------------+
                       \

+-----------+                        +--------------------------+ 
| machine 2 |                        |   machine 5 (replica 2)  |
+-----------+          -             +--------------------------+
|   'b'     |                        |       'c', 'a', 'b'      |
+-----------+                        +--------------------------+
                       /

+-----------+                        +--------------------------+ 
| machine 3 |                        |   machine 5 (replica 2)  |
+-----------+                        +--------------------------+ 
|   'c'     |                        |       'b', 'c', 'a'      |
+-----------+                        +--------------------------+

```

so when insert into the table with unique key, we should let the final instance satisfy the distribution same with the key, and you should add a hash(unique key) require properties to the top of the select

**Comment by sohardforaname** [comment]
run buildall

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 34.22 seconds
 stream load tsv:          423 seconds loaded 74807831229 Bytes, about 168 MB/s
 stream load json:         22 seconds loaded 2358488459 Bytes, about 102 MB/s
 stream load orc:          59 seconds loaded 1101869774 Bytes, about 17 MB/s
 stream load parquet:          30 seconds loaded 861443392 Bytes, about 27 MB/s
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230424072025_clickbench_pr_133873.html

**Comment by sohardforaname** [comment]
run buildall

**Comment by sohardforaname** [comment]
run p0

**Comment by sohardforaname** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/SetRequiredProperties.java [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+                int keysNum = stmtCtx.getInsertIntoContext().getKeyNums();
+                List<ExprId> outputs = plan.getOutput().subList(0, keysNum).stream()
+                        .map(NamedExpression::getExprId).collect(Collectors.toList());
+                cascadesCtx.setJobContext(PhysicalProperties.createHash(
+                        new DistributionSpecHash(outputs, ShuffleType.NATURAL)));
+            }
+        }
+        if (cascadesCtx.getCurrentJobContext().getRequiredProperties() == null) {
+            cascadesCtx.setJobContext(PhysicalProperties.GATHER);
+        }
```
Comment (924060929) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/SetRequiredProperties.java:63]: 1. if is ExplainCommand: you should expend ExplainCommand and according the inner statement to set required properties;
2. if is InsertIntoTableCommand, hash require is right, but I don't know whether ShuffleType.NATURAL is the true require, @morrySnow PTAL
3. you are right, else cases use gather require for query

**Comment by sohardforaname** [comment]
run buildall

**Comment by sohardforaname** [comment]
run buildall

**Comment by sohardforaname** [comment]
run buildall

**Comment by sohardforaname** [comment]
run buildall

**Comment by sohardforaname** [comment]
run fe ut

**Comment by sohardforaname** [comment]
run p0

**Comment by sohardforaname** [comment]
run buildall

**Comment by sohardforaname** [comment]
run buildall

**Comment by sohardforaname** [comment]
run buildall

**Comment by sohardforaname** [comment]
run buildall

**Comment by sohardforaname** [comment]
run buildall

**Comment by sohardforaname** [comment]
run p0

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/commands/InsertIntoTableCommand.java [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+import java.util.stream.Collectors;
+
+/**
+ * insert into select command implementation
+ *
+ * insert into select command support the grammer: explain? insert into table columns? partitions? hints? query
+ * InsertIntoTableCommand is a command to represent insert the answer of a query into a table.
+ * class structure's:
+ *  InsertIntoTableCommand(Query())
+ *  InsertIntoTableCommand(ExplainCommand(Query()))
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/commands/InsertIntoTableCommand.java:66]: ExplainCommand(InsertIntoTableCommand(Query())) is more make sense

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/commands/InsertIntoTableCommand.java [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+    private final List<String> tableName;
+    private final List<String> colNames;
+    private final LogicalPlan logicalQuery;
+    private final String labelName;
+    private Database database;
+    private Table table;
+    private NereidsPlanner planner;
+    private TupleDescriptor olapTuple;
+    private List<String> partitions;
+    private List<String> hints;
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/commands/InsertIntoTableCommand.java:79]: not use?

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/commands/InsertIntoTableCommand.java [hunk 4]
```diff
@@ -0,0 +1,271 @@
...
+    private List<Column> targetColumns;
+    private List<Long> partitionIds = null;
+
+    /**
+     * constructor
+     */
+    public InsertIntoTableCommand(List<String> tableName, String labelName, List<String> colNames,
+            List<String> partitions, List<String> hints, LogicalPlan logicalQuery) {
+        super(PlanType.INSERT_INTO_SELECT_COMMAND);
+        Preconditions.checkArgument(tableName != null, "tableName cannot be null in insert-into-select command");
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/commands/InsertIntoTableCommand.java:89]:
```suggestion
        Preconditions.checkArgument(tableName != null, "tableName cannot be null in InsertIntoTableCommand");
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/PlanType.java [hunk 5]
```diff
@@ -89,5 +89,6 @@ public enum PlanType {
 
     COMMAND,
     EXPLAIN_COMMAND,
-    CREATE_POLICY_COMMAND
+    CREATE_POLICY_COMMAND,
+    INSERT_INTO_SELECT_COMMAND
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/PlanType.java:93]:
```suggestion
    INSERT_INTO_TABLE_COMMAND
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/commands/InsertIntoTableCommand.java [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+    }
+
+    public NereidsPlanner getPlanner() {
+        return planner;
+    }
+
+    @Override
+    public void run(ConnectContext ctx, StmtExecutor executor) throws Exception {
+        if (ctx.isTxnModel()) {
+            // in original planner and in txn model, we can execute sql like: insert into t select 1, 2, 3
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/commands/InsertIntoTableCommand.java:106]:
```suggestion
            // in original planner with txn model, we can execute sql like: insert into t select 1, 2, 3
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/commands/InsertIntoTableCommand.java [hunk 7]
```diff
@@ -0,0 +1,271 @@
...
+            // but no data will be inserted, now we adjust forbid it.
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/commands/InsertIntoTableCommand.java:107]: why no data insert? find the reason and add more info in this comment

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/commands/InsertIntoTableCommand.java [hunk 8]
```diff
@@ -0,0 +1,271 @@
...
+    private TupleDescriptor olapTuple;
+    private List<String> partitions;
+    private List<String> hints;
+    private List<Column> targetColumns;
+    private List<Long> partitionIds = null;
+
+    /**
+     * constructor
+     */
+    public InsertIntoTableCommand(List<String> tableName, String labelName, List<String> colNames,
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/commands/InsertIntoTableCommand.java:86]:
```suggestion
    public InsertIntoTableCommand(List<String> nameParts, String labelName, List<String> colNames,
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/commands/InsertIntoTableCommand.java [hunk 9]
```diff
@@ -0,0 +1,271 @@
...
+        if (isExplain()) {
+            executor.handleExplainStmt(((ExplainCommand) logicalQuery).getExplainString(planner));
+            return;
+        }
+
+        txn.executeInsertIntoSelectCommand(executor);
+    }
+
+    private void checkDatabaseAndTable(ConnectContext ctx) {
+        List<String> qualifier = RelationUtil.getQualifierName(ctx, tableName);
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/commands/InsertIntoTableCommand.java:154]:
```suggestion
        List<String> qualifiedTableName = RelationUtil.getQualifierName(ctx, tableName);
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/commands/InsertIntoTableCommand.java [hunk 10]
```diff
@@ -0,0 +1,271 @@
...
+
+        if (isExplain()) {
+            executor.handleExplainStmt(((ExplainCommand) logicalQuery).getExplainString(planner));
+            return;
+        }
+
+        txn.executeInsertIntoSelectCommand(executor);
+    }
+
+    private void checkDatabaseAndTable(ConnectContext ctx) {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/commands/InsertIntoTableCommand.java:153]: bindTargetRelation

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/commands/InsertIntoTableCommand.java [hunk 11]
```diff
@@ -0,0 +1,271 @@
...
+        List<String> qualifier = RelationUtil.getQualifierName(ctx, tableName);
+        String catalogName = qualifier.get(0);
+        String dbName = qualifier.get(1);
+        String tableName = qualifier.get(2);
+        CatalogIf catalog = Env.getCurrentEnv().getCatalogMgr().getCatalog(catalogName);
+        if (catalog == null) {
+            throw new RuntimeException(String.format("Catalog %s does not exist.", catalogName));
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/commands/InsertIntoTableCommand.java:160]:
```suggestion
            throw new AnalysisException(String.format("Catalog %s does not exist.", catalogName));
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/commands/InsertIntoTableCommand.java [hunk 12]
```diff
@@ -0,0 +1,271 @@
...
+        }
+        try {
+            database = ((Database) catalog.getDb(dbName).orElseThrow(() ->
+                    new RuntimeException("Database [" + dbName + "] does not exist.")));
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/commands/InsertIntoTableCommand.java:164]:
```suggestion
                    new AnalysisException("Database [" + dbName + "] does not exist.")));
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/commands/InsertIntoTableCommand.java [hunk 13]
```diff
@@ -0,0 +1,271 @@
...
+            table = database.getTable(tableName).orElseThrow(() ->
+                    new RuntimeException("Table [" + tableName + "] does not exist in database [" + dbName + "]."));
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/commands/InsertIntoTableCommand.java:166]:
```suggestion
                    new AnalysisException("Table [" + tableName + "] does not exist in database [" + dbName + "]."));
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/commands/InsertIntoTableCommand.java [hunk 14]
```diff
@@ -0,0 +1,271 @@
...
+    @Override
+    public void run(ConnectContext ctx, StmtExecutor executor) throws Exception {
+        if (ctx.isTxnModel()) {
+            // in original planner and in txn model, we can execute sql like: insert into t select 1, 2, 3
+            // but no data will be inserted, now we adjust forbid it.
+            throw new AnalysisException("insert into table command is not supported in txn model");
+        }
+        checkDatabaseAndTable(ctx);
+        getColumns();
+        getPartition();
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/commands/InsertIntoTableCommand.java:112]: put the code block back to `run` method and remove these functions

File: fe/fe-core/src/main/java/org/apache/doris/nereids/txn/Transaction.java [hunk 15]
```diff
@@ -0,0 +1,243 @@
...
+import com.google.common.base.Strings;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Lists;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+/**
+ * transaction wrapper for Nereids
+ */
+public class Transaction {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/txn/Transaction.java:55]: why need Nereids' transaction?

File: fe/fe-core/src/main/java/org/apache/doris/nereids/NereidsPlanner.java [hunk 16]
```diff
@@ -209,6 +207,8 @@ public Plan plan(LogicalPlan plan, PhysicalProperties requireProperties, Explain
 
             deriveStats();
 
+            adjustRequiredProperties(cascadesContext.getRewritePlan());
```
[the remark this answers is not in the document]
Comment (sohardforaname) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/NereidsPlanner.java:210]: because we may handle star which will be expanded in analyzer.

File: fe/fe-core/src/main/java/org/apache/doris/nereids/txn/Transaction.java (hunk 15 above)
Comment (sohardforaname) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/txn/Transaction.java:55]: because the handleInsertStmt() in original planner is too hard to be rewritten

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/commands/InsertIntoTableCommand.java (hunk 10 above)
Comment (sohardforaname) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/commands/InsertIntoTableCommand.java:153]: What does it mean？

**Comment by sohardforaname** [comment]
run buildall

**Comment by sohardforaname** [comment]
run buildall

**Comment by sohardforaname** [comment]
run p0

**Comment by sohardforaname** [comment]
run p0

**Comment by sohardforaname** [comment]
run p0

**Comment by sohardforaname** [comment]
run buildall

**Comment by sohardforaname** [comment]
run p0

**Comment by sohardforaname** [comment]
run p0

**Comment by sohardforaname** [comment]
run p0

**Comment by sohardforaname** [comment]
run p0

**Comment by sohardforaname** [comment]
run buildall

**Comment by sohardforaname** [comment]
run p0

**Comment by sohardforaname** [comment]
run p0

**Comment by morrySnow** [comment]
run buildall

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 34.3 seconds
 stream load tsv:          418 seconds loaded 74807831229 Bytes, about 170 MB/s
 stream load json:         22 seconds loaded 2358488459 Bytes, about 102 MB/s
 stream load orc:          60 seconds loaded 1101869774 Bytes, about 17 MB/s
 stream load parquet:          30 seconds loaded 861443392 Bytes, about 27 MB/s
 insert into select:          78.4 seconds inserted 10000000 Rows, about 127K ops/s
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230523031506_clickbench_pr_148616.html

**Comment by sohardforaname** [comment]
run buildall

**Comment by sohardforaname** [comment]
run p0

**Comment by sohardforaname** [comment]
run p0

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/commands/InsertIntoTableCommand.java [hunk 17]
```diff
@@ -0,0 +1,140 @@
...
+/**
+ * insert into select command implementation
+ * insert into select command support the grammer: explain? insert into table columns? partitions? hints? query
+ * InsertIntoTableCommand is a command to represent insert the answer of a query into a table.
+ * class structure's:
+ *  InsertIntoTableCommand(Query())
+ *  ExplainCommand(Query())
+ */
+public class InsertIntoTableCommand extends Command implements ForwardWithSync {
+    public static final Logger LOG = LogManager.getLogger(InsertIntoTableCommand.class);
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/commands/InsertIntoTableCommand.java:51]: add a blank line before this line

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/commands/InsertIntoTableCommand.java [hunk 18]
```diff
@@ -0,0 +1,140 @@
...
+    private final String labelName;
+    private NereidsPlanner planner;
+    private boolean isTxnBegin = false;
+
+    /**
+     * constructor
+     */
+    public InsertIntoTableCommand(LogicalPlan logicalQuery, String labelName) {
+        super(PlanType.INSERT_INTO_TABLE_COMMAND);
+        Preconditions.checkArgument(logicalQuery != null, "logicalQuery cannot be null in InsertIntoTableCommand");
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/commands/InsertIntoTableCommand.java:63]: Preconditions.checkNotNull

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/commands/ExplainCommand.java
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/commands/ExplainCommand.java:1]: revert these changes

**Comment by sohardforaname** [comment]
run buildall

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 34.58 seconds
 stream load tsv:          415 seconds loaded 74807831229 Bytes, about 171 MB/s
 stream load json:         22 seconds loaded 2358488459 Bytes, about 102 MB/s
 stream load orc:          59 seconds loaded 1101869774 Bytes, about 17 MB/s
 stream load parquet:          30 seconds loaded 861443392 Bytes, about 27 MB/s
 insert into select:          78.4 seconds inserted 10000000 Rows, about 127K ops/s
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230523091037_clickbench_pr_148848.html

**Comment by sohardforaname** [comment]
run buildall

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 34.73 seconds
 stream load tsv:          419 seconds loaded 74807831229 Bytes, about 170 MB/s
 stream load json:         22 seconds loaded 2358488459 Bytes, about 102 MB/s
 stream load orc:          59 seconds loaded 1101869774 Bytes, about 17 MB/s
 stream load parquet:          30 seconds loaded 861443392 Bytes, about 27 MB/s
 insert into select:          77.6 seconds inserted 10000000 Rows, about 128K ops/s
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230523122834_clickbench_pr_149042.html

**Comment by sohardforaname** [comment]
run buildall

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 34.51 seconds
 stream load tsv:          414 seconds loaded 74807831229 Bytes, about 172 MB/s
 stream load json:         21 seconds loaded 2358488459 Bytes, about 107 MB/s
 stream load orc:          59 seconds loaded 1101869774 Bytes, about 17 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          78.5 seconds inserted 10000000 Rows, about 127K ops/s
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230523130232_clickbench_pr_149053.html

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 34.04 seconds
 stream load tsv:          417 seconds loaded 74807831229 Bytes, about 171 MB/s
 stream load json:         22 seconds loaded 2358488459 Bytes, about 102 MB/s
 stream load orc:          60 seconds loaded 1101869774 Bytes, about 17 MB/s
 stream load parquet:          30 seconds loaded 861443392 Bytes, about 27 MB/s
 insert into select:          78.0 seconds inserted 10000000 Rows, about 128K ops/s
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230523132539_clickbench_pr_149083.html

**Comment by sohardforaname** [comment]
run buildall

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 34.89 seconds
 stream load tsv:          420 seconds loaded 74807831229 Bytes, about 169 MB/s
 stream load json:         22 seconds loaded 2358488459 Bytes, about 102 MB/s
 stream load orc:          59 seconds loaded 1101869774 Bytes, about 17 MB/s
 stream load parquet:          30 seconds loaded 861443392 Bytes, about 27 MB/s
 insert into select:          79.1 seconds inserted 10000000 Rows, about 126K ops/s
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230524072841_clickbench_pr_149576.html

**Comment by sohardforaname** [comment]
run buildall

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 34.8 seconds
 stream load tsv:          416 seconds loaded 74807831229 Bytes, about 171 MB/s
 stream load json:         22 seconds loaded 2358488459 Bytes, about 102 MB/s
 stream load orc:          60 seconds loaded 1101869774 Bytes, about 17 MB/s
 stream load parquet:          30 seconds loaded 861443392 Bytes, about 27 MB/s
 insert into select:          79.0 seconds inserted 10000000 Rows, about 126K ops/s
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230524093627_clickbench_pr_149652.html

File: fe/fe-core/src/main/java/org/apache/doris/nereids/NereidsPlanner.java [hunk 19]
```diff
@@ -95,6 +93,7 @@ public class NereidsPlanner extends Planner {
     private Plan analyzedPlan;
     private Plan rewrittenPlan;
     private Plan optimizedPlan;
+    private PhysicalPlan physicalPlan;
```
[the remark this answers is not in the document]
Comment (sohardforaname) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/NereidsPlanner.java:96]: no, will be used in insert into table command

**Comment by sohardforaname** [comment]
run buildall

**Review by morrySnow — APPROVED** [review]

**Comment by morrySnow** [comment]
do not merge this PR until the author update commit msg
